nanostack-event-loop/eventOS_event.h

```diff
@@ -19,6 +19,173 @@
+ *
+ * @startuml
+ *      partition tasklet1.cpp {
+ *          (*) --> tasklet1
+ *      }
+ *      partition "eventOS" {
+ *          tasklet1 -->[event:\nreceiver: 3\nevent_id: 1] eventOS_event_send
+ *      }
+ *      partition tasklet3.cpp {
+ *          eventOS_event_send -->[event:\nreceiver: 3\nevent_id: 1] tasklet3
+ *      }
+ * @enduml
+ *
+ * <h3>Usage</h3>
+ *
+ * To send or receive events, you first need to register your event handler.
+ * \code
+ *      // In header
+ *      extern uint8_t my_eventhandler_id;
+ *      #define INITIALIZATION_EVENT 0
+ *      #define MY_EVENT 1
+ *
+ *      // In my_handler.cpp
+ *      void my_event_handler(arm_event_t *e)
+ *      {
+ *          switch (e->event_type) {
+ *              case INITIALIZATION_EVENT:
+ *                  // Initialize my module
+ *                  break;
+ *              case MY_EVENT:
+ *                  // Event received
+ *                  break;
+ *          }
+ *      }
+ *
+ *      // Register the handler
+ *      my_eventhandler_id = eventOS_event_handler_create(my_event_handler, INITIALIZATION_EVENT);
+ *      if (my_eventhandler_id < 0) {
+ *          // fail
+ *      }
+ * \endcode
+ *
+ * Each event is basically a \ref arm_event_s structure. You need to fill in the arm_event_s::receiver field.
+ * Rest of the fields are optional, and used only by the receiving callback. So you have different options to
+ * deliver data to a receiving tasklet. The structure is copied by the event system, so temporary storage may be used,
+ * and the structure may be freed after it has been pushed into event system.
+ *
+ * \code
+ *      // Send the event
+ *      arm_event_t e = {
+ *       .receiver = my_eventhandler_id,
+ *       .event_type = MY_EVENT
+ *      };
+ *
+ *      if (eventOS_event_send(e) != 0) {
+ *       // fail
+ *      }
+ * \endcode
+ *
+ * Where required, event system allows you to delay the event propagation.
+ *
+ * \code
+ *      // Wait 3 seconds before the event
+ *      #define MY_DELAY_MS 3000
+ *
+ *      arm_event_t e = {
+ *          .receiver = my_eventhandler_id,
+ *          .event_type = MY_EVENT
+ *      };
+ *
+ *      uint32_t delay = eventOS_event_timer_ms_to_ticks(MY_DELAY_MS);
+ *      eventOS_event_send_after(e, delay);
+ * \endcode
+ *
+ * \sa eventOS_event.h
+ * \sa eventOS_event_send_at
+ * \sa eventOS_event_send_in
+ * \sa eventOS_event_send_after
+ * \sa eventOS_event_send_every
+ *
+ * <h3>Pre-allocated events</h3>
+ *
+ * Two options are provided to limit the heap usage. First option is to use reccuring events with eventOS_event_send_every(),
+ * so your event is only allocated once. This allows you to create application that does not use heap after initialization phase.
+ *
+ * Second option is to use pre-allocated or statically allocated event structure. In this model you create a space for
+ * \ref arm_event_storage structure and send events using eventOS_event_send_user_allocated() call.
```

Comment:
This might even encourage one to prefer the pre-allocated events as their sending can not fail.

Reply:
I'll ad that as well.